### PR TITLE
chore: enable noImplicitAny

### LINKED ISSUE #69: Enable noImplicitAny in the tsconfig

### PR DESCRIPTION
**Pull request by JustinBeckwith**
This was fairly painful.  It ain't perfect, but it gives us a good starting point to improve the typing now as we move along.  Resolves #69.

### CODE REVIEW

File: test/util.ts [hunk 1]
```diff
@@ -332,7 +337,7 @@ describe('common/util', () => {
     });
 
     it('should link the original interceptors_', () => {
-      const interceptors = [];
+      const interceptors = new Array<{}>();
```
Comment (kjin) [inline test/util.ts]: Also here for array constructor comment.

File: src/util.ts [hunk 2]
```diff
@@ -142,12 +142,22 @@ export interface OnAuthenticatedCallback {
...
 }
 
+export class GoogleError extends Error {
+  code?: number;
+  errors?: Array<GoogleInnerError>;
+  response?: r.Response;
+}
 
-export interface GoogleError extends Error {
+export interface GoogleErrorBody {
```
[the remark this answers is not in the document]
Comment (JustinBeckwith) [reply src/util.ts]: I was able to just drop `GoogleError` and use `ApiError` in most cases.

File: src/util.ts [hunk 3]
```diff
@@ -520,8 +545,8 @@ export class Util {
         if (!err || autoAuthFailed) {
           let projectId = authClient.projectId;
 
-          if (config.projectId && config.projectId !== '{{projectId}}') {
-            projectId = config.projectId;
+          if (config!.projectId && config!.projectId !== '{{projectId}}') {
```
[the remark this answers is not in the document]
Comment (JustinBeckwith) [reply src/util.ts]: I just changed the function signature to do a default assignment if the value wasn't provided

File: src/util.ts [hunk 4]
```diff
@@ -592,14 +617,10 @@ export class Util {
...
     }
-
-    (makeAuthenticatedRequest as MakeAuthenticatedRequest).getCredentials = authClient.getCredentials.bind(
-      authClient
-    );
-
-    (makeAuthenticatedRequest as MakeAuthenticatedRequest).authClient = authClient;
-
-    return makeAuthenticatedRequest;
+    const mar = makeAuthenticatedRequest as MakeAuthenticatedRequest;
```
[the remark this answers is not in the document]
Comment (JustinBeckwith) [reply src/util.ts]: Ahah!  I thought so too initially.  The problem is that you can't use function overloading when you do it as a variable assignment.  The tradeoff here is between expressing the whack function signature (which is indeed, quite whack) vs the cast. 

Given the fact that a ton of this code will get simplified when we go async/await, I'd prefer to keep this as is for now.

File: test/operation.ts [hunk 5]
```diff
@@ -201,11 +201,11 @@ describe('Operation', () => {
           error: {},
         };
 
-        operation.getMetadata = (callback) => {
+        operation.getMetadata = (callback: (err: null, a: {}, b: {}) => void) => {
```
[the remark this answers is not in the document]
Comment (JustinBeckwith) [reply test/operation.ts]: Good call!

File: test/paginator.ts [hunk 6]
```diff
@@ -100,7 +100,7 @@ describe('paginator', () => {
         return this.uuid;
       };
 
-      const cls = new FakeClass();
+      const cls = new (FakeClass as any)();
```
[the remark this answers is not in the document]
Comment (JustinBeckwith) [reply test/paginator.ts]: I tried this initially.  The problem is that the rest of test arbitrarily pins new methods on the prototype :|

File: test/service-object.ts [hunk 7]
```diff
@@ -657,7 +664,7 @@ describe('ServiceObject', () => {
         reqOpts.uri,
       ].join('/');
 
-      (serviceObject as any).parent.request = function(reqOpts_, callback) {
+      (serviceObject as any).parent.request = function(reqOpts_: ExtendedRequestOptions, callback: () => void) {
```
[the remark this answers is not in the document]
Comment (JustinBeckwith) [reply test/service-object.ts]: Sadly the TypeScript compiler was too smart for this trick.  It noticed that I was adding a property that was private on the class, and threw an error.  The real fix here is to write tests that aren't dependent on the internals (private members) of the class.  I was able to hide some of the jank behind a method, but it's still jank indeed.

**Pull request merged**